presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/GroupTraitCalculator.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.facebook.presto.sql.planner.iterative;
+
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.List;
+
+public interface GroupTraitCalculator<T extends GroupTrait>
```

Comment:
> I don't think dedicated GroupTraitCalculator (or TraitCalculator for that matter) is required.

Sounds good.

> I think we should focus first on how to represent traits for individual plan nodes and how to incorporate trait computation into rule processing (in a pluggable/non-static way). It seems that computing traits and applying rules is very similar. Because of that, I think traits should be computed as rules. This way we can have pluggable traits rules and also inspect/add traits during rule execution. Having this separation: trait calculator/rules seems artificial and unnecessary.

Sounds like an overkill. Trait calculations is very easy, having rules for each node will blow up this simple class into several rules.